### ISSUE TITLE
global: fully functional PostgreSQL DB backend

### ISSUE THREAD

**Issue by hachreak** [inveniosoftware/invenio#3079]
The list of issues and PR that we need to solve before we can think to add PostgreSQL:

- [x] global: PostgreSQL Text type compatibility fix https://github.com/inveniosoftware/invenio/pull/3045
- [x] sqlalchemy: PostgreSQL - type Text and Integer compatibility https://github.com/inveniosoftware/invenio/issues/3037
- [x] general: fix models for PostgreSQL compatibility https://github.com/inveniosoftware/invenio/pull/2899
- [ ] equivalent function [get_table_update_time()](https://github.com/inveniosoftware/invenio/blob/master/invenio/legacy/dbquery_mysql.py#L439) and [get_table_status_info()](https://github.com/inveniosoftware/invenio/blob/master/invenio/legacy/dbquery_mysql.py#L471) for postgresql. (@jacquerie) #3211
- [x] bibsort: fix `database create` error (`bibsort -u admin --load-config` return error) #3212
- [ ] travis: add PostgreSQL to test matrix #3213
- [ ] legacy: remove mysql specific codes (see `git grep MySQLdb`, `git grep "SHOW TABLES"`, `git grep "CHECK TABLE"`, `git grep "RLIKE"`, `git grep "PRIORITY"`, `git grep DATE_FORMAT`, `regexp`, `git grep "ON DUPLICATE KEY UPDATE"`) #3214
- [x] DB run_sql problems https://github.com/inveniosoftware/invenio/issues/3089 (legacy: tables name and columns name that have upper case on sql queries require double quote, values in query require single quote, table `user` require double quote)
- [x]  legacy: fix run_sql ansi quotes in bibsched https://github.com/inveniosoftware/invenio/pull/3093
- [x] legacy: fix run_sql ansi quotes https://github.com/inveniosoftware/invenio/pull/3096
- [x] access: fix foreign key declaration. #3099
- [x] oauth2server: oauth2CLIENT not cleaned if user is deleted #3145 
- [x] SQLAlchemy: boolean value on PostgreSQL https://github.com/inveniosoftware/invenio/issues/3215
- [x]  PostgreSQL: serial id is not update after manually creation a admin user https://github.com/inveniosoftware/invenio/issues/3216
- [x]  SQLAlchemy: UUID value on PostgreSQL https://github.com/inveniosoftware/invenio/issues/3217
- [ ]  PostgreSQL: foreign key problems when delete records https://github.com/inveniosoftware/invenio/issues/3218
- [x]  PostgreSQL: use a passlib to create cookie instead of internal AES function https://github.com/inveniosoftware/invenio/issues/3219
- [x]  tags: error if it assignes value 0 on foreign keys https://github.com/inveniosoftware/invenio-tags/issues/3
- [x] access: cookie generation fail with PostgreSQL #3318
- [x] accounts: user creation with id = 1 problem on PostgreSQL #3320

--
- related #3098

**Comment by jacquerie** [comment]
Hey @hachreak!

Are you working on
* equivalent function get_table_update_time() and get_table_status_info() for postgresql
?

If not, I'd like to tackle it.

**Comment by hachreak** [comment]
Have you some idea how do easily?

**Comment by jacquerie** [comment]
No, not really. In fact, from what I see, there's no way to do it easily: http://stackoverflow.com/a/12755522/374865. The other thing I'm seeing is that `get_table_status_info` only returns the status information from the last table it selected: https://github.com/inveniosoftware/invenio/blob/e9ff05f713c2e43bf8b2da1771868a5f1dc0e063/invenio/legacy/dbquery_mysql.py#L481-L511, probably not what was intended.

What I really hope is that these functions are not needed, and the logic they provide can be implemented in an easier way (this seems to be the case in https://github.com/inveniosoftware/invenio/blob/e9ff05f713c2e43bf8b2da1771868a5f1dc0e063/invenio/legacy/bibindex/adminlib.py#L2642-L2733).

**Comment by jirikuncar** [comment]
@jacquerie can you try to replace the usage of `get_table_update_time` by something like `MAX(update_time)` if exists in the table?

**Comment by jacquerie** [comment]
Unfortunately, not all tables on which the `get_table_update_time` function is called have an `update_time` field, and when they a similar field it might have a different name.

For example:
```SQL
mysql> show columns from sbmDOCTYPE;
+-------------+-------------+------+-----+---------+-------+
| Field       | Type        | Null | Key | Default | Extra |
+-------------+-------------+------+-----+---------+-------+
| ldocname    | text        | YES  |     | NULL    |       |
| sdocname    | varchar(10) | NO   | PRI |         |       |
| cd          | date        | YES  |     | NULL    |       |
| md          | date        | YES  |     | NULL    |       |
| description | text        | YES  |     | NULL    |       |
+-------------+-------------+------+-----+---------+-------+
5 rows in set (0.03 sec)
```
where I'm guessing that `md` stands for `modification_date`. So one could write a function that uses the right column name and uses `MAX` on that, but it looks like a terrible hack on top of another hack.

On the bright side, I don't think we need to care about `get_table_status_info`: the only places it's used are in `bibindex`, which is going to be killed anyway by ElasticSearch:
```shell
(invenio2) ~/.virtualenvs/invenio2/src/invenio$ ag --files-with-matches get_table_status_info
invenio/legacy/dbquery_postgresql.py
invenio/legacy/dbquery_mysql.py
invenio/legacy/bibindex/scripts/bibstat.py
invenio/legacy/bibindex/adminlib.py
```

**Comment by jacquerie** [comment]
@hachreak: I'm also having a look at
* legacy: remove mysql specific codes (see git grep MySQLdb, git grep "SHOW TABLES", git grep "CHECK TABLE", git grep "RLIKE", git grep "PRIORITY", git grep DATE_FORMAT, regexp)

In particular, `CHECK TABLE` is only used once in `invenio/legacy/websession/inveniogc.py` as:
```python
def check_tables():
    """
    Check all DB tables.  Useful to run from time to time when the
    site is idle, say once a month during a weekend night.

    FIXME: should produce useful output about outcome.
    """
    res = run_sql("SHOW TABLES")
    for row in res:
        table_name = row[0]
        write_message("checking table %s" % table_name)
        run_sql("CHECK TABLE %s" % wash_table_column_name(table_name)) # kwalitee: disable=sql
```

I think this can be safely dropped. In fact, what about dropping the entire `inveniogc.py`? Is someone using it?

**Comment by jirikuncar** [comment]
> In fact, what about dropping the entire `inveniogc.py`? Is someone using it?

@jacquerie Create separate RFC. IMHO it is still used.

**Comment by tiborsimko** [comment]
Note that I've created a separate issue for every item listed in the original ticket description that was not having one yet.  

In this way we can discuss specific points in their own issues, and keep this meta-issue very slim here, basically for overview purposes.  If something new pops out, let's always open a new issue and let's reference it in the original description here. In this way people can always see the list of our active bigger roadmap tasks (=PostgreSQL "work package"), what is their status (=5 out of 22 subtasks completed), for which milestone they are targeted (=2.2), etc, via the following URL:

- https://github.com/inveniosoftware/invenio/issues?q=label:roadmap

In this way, our bigger roadmap tasks (=move to PostgreSQL) can also be more easily split among volunteering people, e.g. John takes care of T1, Jane takes care of T2, while Jim maintains the overall overview status in this meta-issue. 

P.S. I've started to populate some of the roadmap tasks in this way, and more are to come according to our common "Invenio 3" roadmap vision.

**Comment by jacquerie** [comment]
Yet another (small) issue: we can't pass a `LocalProxy` as an argument to a query. To see the full error run `$ py.test invenio/testsuite/test_restful.py` as there's lots of output. Now, this problem can be solved by converting the content of the `LocalProxy` to the right type, but I'm not sure if that is the right solution.

**Comment by tiborsimko** [comment]
PostgreSQL is fully supported in Invenio 3.
